tests/system/conftest.py

```diff
@@ -92,10 +94,237 @@ def numeric_table(client):
     return client.table('numeric_table')
 
 
+@pytest.fixture(scope='session')
+def df(alltypes):
+    return alltypes.execute()
+
+
+@pytest.fixture(scope='session')
+def parted_df(parted_alltypes):
+    return parted_alltypes.execute()
+
+
 @pytest.fixture(scope='session')
 def public(project_id, credentials):
     return bq.connect(
         project_id=project_id,
         dataset_id='bigquery-public-data.stackoverflow',
         credentials=credentials,
     )
+
+
+# Native BigQuery client fixtures
+# required to dynamically create the testing dataset,
+# the tables, and to populate data into the tables.
+@pytest.fixture(scope='session')
+def bqclient(client):
+    return client.client
+
+
+# Create testing dataset.
+@pytest.fixture(scope='session')
+def testing_dataset(bqclient, request):
+    dataset_ref = bigquery.DatasetReference(bqclient.project, DATASET_ID)
+    overwrite = request.config.getoption("--overwrite")
+    delete_dataset = False
+    try:
+        bqclient.get_dataset(dataset_ref)
+        if overwrite:
+            delete_dataset = True
+        else:
+            pytest.exit(
+                'Testing dataset exists, use --overwrite option to replace.')
+    except NotFound:
+        pass
+    if delete_dataset:
+        bqclient.delete_dataset(dataset_ref, delete_contents=True)
+    bqclient.create_dataset(dataset_ref)
+    yield dataset_ref
+    bqclient.delete_dataset(dataset_ref, delete_contents=True)
```

Comment:
We'll have to check what our minimum `google-cloud-bigquery` version is, but I believe we can simplify this logic (avoiding the call to `get_dataset`).

* If overwrite, then call [`delete_datatset`](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.delete_dataset) with `not_found_ok=True`
* Always, call [`create_dataset`](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.create_dataset) with `exists_ok=True`.

Reply:
`google-cloud-bigquery >=1.12.0` is the minimum install requirement, `exists_ok` was added in version 1.10, so we should be ok.  However, `exists_ok` on the dataset leaves any existing tables and data in place, and this can result in failing test, mainly because of an inconsistent use of `write_disposition`.

I wanted the option to persevere the data in the testing tables to debug why something not be working.  So, I think the strategy should be to change my option from `--overwrite` to `--save-dataset` so that the dataset can be preserved for inspection.  And, then set the `write_disposition` to `WRITE_TRUNCATE` to ensure that sequential testing doesn't fail.

Reply:
Sounds good. I've also added a `--no-refresh-dataset` option to speed up tests when I was debugging locally.